### ISSUE TITLE
[BUG] VM backup cause harvester pod to crash

### ISSUE THREAD

**Issue by guangbochen** [harvester/harvester#1588]
**Describe the bug**
VM backup cause harvester pod to crash after upgrade

**To Reproduce**
1. create a VM backup with v0.3.0 cluster
1. upgrade harvester from v0.3.0 to the master-head version
1. the harvester pod keeps crash

**Additional context**
```
time="2021-11-26T02:32:10Z" level=error msg="error syncing 'test/yz-template-test-8v6jx': handler template-version-controller: virtualmachinetemplates.harvesterhci.io \"yz-template-test\" not found, requeuing"
E1126 02:32:10.964692       7 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1018 [running]:
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2147040, 0x3b3a920)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x95
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x86
panic(0x2147040, 0x3b3a920)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/harvester/harvester/pkg/controller/master/backup.IsBackupTargetSame(...)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/util.go:31
github.com/harvester/harvester/pkg/controller/master/backup.(*Handler).uploadVMBackupMetadata(0xc00396c900, 0xc0000ec580, 0xc0005c5200, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/backup.go:564 +0x6c
github.com/harvester/harvester/pkg/controller/master/backup.(*Handler).OnBackupChange(0xc00396c900, 0xc000a82130, 0xb, 0xc0000ec580, 0xd50be1e5edfc61b6, 0x40409a, 0x403bcb)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/backup.go:120 +0x39f
github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1.FromVirtualMachineBackupHandlerToHandler.func1(0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0x756ea1c39c62ca, 0x1772834797ccb030, 0x403bcb, 0xc00f32c2b0)
	/go/src/github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1/virtualmachinebackup.go:102 +0x6b
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc00f32c2a0, 0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0xc00f32c2b0, 0x756ea14cc4fc4b, 0x40a63f, 0xc00003a000)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/sharedcontroller.go:29 +0x4e
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc0003bf100, 0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0xc003030501, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/sharedhandler.go:69 +0x14c
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc000b6ebb0, 0xc000a82130, 0xb, 0xc0005cd798, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:215 +0xd1
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc000b6ebb0, 0x2027ca0, 0xc00f32c2b0, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:197 +0xe7
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc000b6ebb0, 0x203001)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:174 +0x54
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:163
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0021675a0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0021675a0, 0x27f0340, 0xc000909a40, 0x75676e69733a6601, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x9b
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0021675a0, 0x3b9aca00, 0x0, 0x3a227d225c646501, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0021675a0, 0x3b9aca00, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:134 +0x33b
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1e45bcc]

goroutine 1018 [running]:
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x109
panic(0x2147040, 0x3b3a920)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/harvester/harvester/pkg/controller/master/backup.IsBackupTargetSame(...)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/util.go:31
github.com/harvester/harvester/pkg/controller/master/backup.(*Handler).uploadVMBackupMetadata(0xc00396c900, 0xc0000ec580, 0xc0005c5200, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/backup.go:564 +0x6c
github.com/harvester/harvester/pkg/controller/master/backup.(*Handler).OnBackupChange(0xc00396c900, 0xc000a82130, 0xb, 0xc0000ec580, 0xd50be1e5edfc61b6, 0x40409a, 0x403bcb)
	/go/src/github.com/harvester/harvester/pkg/controller/master/backup/backup.go:120 +0x39f
github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1.FromVirtualMachineBackupHandlerToHandler.func1(0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0x756ea1c39c62ca, 0x1772834797ccb030, 0x403bcb, 0xc00f32c2b0)
	/go/src/github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1/virtualmachinebackup.go:102 +0x6b
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc00f32c2a0, 0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0xc00f32c2b0, 0x756ea14cc4fc4b, 0x40a63f, 0xc00003a000)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/sharedcontroller.go:29 +0x4e
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc0003bf100, 0xc000a82130, 0xb, 0x2814a80, 0xc0000ec580, 0xc003030501, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/sharedhandler.go:69 +0x14c
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc000b6ebb0, 0xc000a82130, 0xb, 0xc0005cd798, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:215 +0xd1
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc000b6ebb0, 0x2027ca0, 0xc00f32c2b0, 0x0, 0x0)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:197 +0xe7
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc000b6ebb0, 0x203001)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:174 +0x54
github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:163
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0021675a0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0021675a0, 0x27f0340, 0xc000909a40, 0x75676e69733a6601, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x9b
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0021675a0, 0x3b9aca00, 0x0, 0x3a227d225c646501, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0021675a0, 0x3b9aca00, 0xc0005bade0)
	/go/src/github.com/harvester/harvester/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/src/github.com/harvester/harvester/vendor/github.com/rancher/lasso/pkg/controller/controller.go:134 +0x33b
```

**Comment by FrankYang0529** [comment]
## Test Plan

* Create a VM Backup with v0.3.0 cluster
* Upgrade VM Backup CRD. Add `longhornBackupName` to `volumeBackups`. Add `secretBackups` and `backupTarget` to `status`.
  * https://github.com/harvester/harvester/pull/1447/commits/cb02453bf9aefdb06313438f09376a7e50c52569#diff-91562ef2708a3a87b91f4fc1ccefc8c95954eb8ca2949f1b76e5f16cc40273ce
  * https://github.com/harvester/harvester/pull/1487/commits/4345ab61f6154439f18cfabca816e108ba7d0626#diff-91562ef2708a3a87b91f4fc1ccefc8c95954eb8ca2949f1b76e5f16cc40273ce
![Screen Shot 2021-11-29 at 11 33 02 AM](https://user-images.githubusercontent.com/4927361/143804598-15904611-1ffc-4164-894e-5cef78dc32c1.png)
![Screen Shot 2021-11-29 at 11 33 17 AM](https://user-images.githubusercontent.com/4927361/143804610-9228991d-ddfc-426d-b9b8-130f878d743e.png)

* Upgrade harvester image from v0.3.0 to the master-head version.
  * Check backup target information moved from annotation to status.
  * Check `LonghornBackupName` is in `VolumeBackups`.
  * Check `<namespace>-<name>.cfg` is in remote backup target.
  * Check information in `<namespace>-<name>.cfg` is same as VM Backup resources.
* Upgrade harvester-webhook image from v0.3.0 to the master-head version and grant permission to harvester-webhook like the following.
  * https://github.com/harvester/harvester/pull/1447/commits/2c1b9c2c062dbad5e59df48af675182da1db7a7c#diff-cbaabc6d4bb79d51b3ac588ab97ffb29c95c69ec8b1d49c00ff8544353459b0e
![Screen Shot 2021-11-29 at 11 20 09 AM](https://user-images.githubusercontent.com/4927361/143803544-c7314f3f-b487-4d78-a576-95fa2af506f6.png)
* Restore a VM from VM Backup.
  *  Check VM can be started and data is same.

**Comment by joshimoo** [comment]
The panic is caused by https://github.com/harvester/harvester/blob/v1.0/pkg/controller/master/backup/backup.go#L566-L574
not having a nil check against
`vmBackup.Status.BackupTarget` 

Since in the case where the annotation is missing and status does not have a `BackupTarget`, you will run into the panic.
The reason the status is not updated correctly is that update calls for a resource that has a status subresource, will ignore the updates to the status sub resource.
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource

This happens here, refs:
https://github.com/harvester/harvester/blob/383345f3740e1a25be6094c0154d9fe31afecf04/pkg/controller/master/backup/backup.go#L711-L720

https://github.com/harvester/harvester/blob/383345f3740e1a25be6094c0154d9fe31afecf04/pkg/generated/controllers/harvesterhci.io/v1beta1/virtualmachinebackup.go#L182-L185

Not sure why there is no preview for this link.
https://github.com/rancher/lasso/blob/6c6cd7fd6607e280ef3addec9901a8f4dbb12295/pkg/client/client.go#L159-L167

Be aware the CRD has the same difference between Update/UpdateStatus
https://github.com/harvester/harvester/blob/383345f3740e1a25be6094c0154d9fe31afecf04/pkg/generated/clientset/versioned/typed/harvesterhci.io/v1beta1/virtualmachinebackup.go#L126

The code will remove the annotation, but ignore the status update.
Which means there will no longer be an annotation as well as no `Status.BackupTarget`

cc @guangbochen This seems to happen in multiple places inside of the backup handling. Where `client.update` is called but actually `updateStatus` should have been called.

I might have missed something though, since i looked at this form the browser and there is a lot of auto generated code which is pretty cool :)

**Comment by weihanglo** [comment]
Note that there is a kubebuilder mark comment that enables subresource
https://book.kubebuilder.io/reference/generating-crd.html#status

However, I cannot inspect why the status field not retrieved.

**Comment by TachunLin** [comment]
Verified **fixed** after upgrade from v0.3.0 to `master-935a1670-head`.(12/17 based on v1.0.0-rc1) 
Close this issue.

### Result
After manual upgrade from v0.3.0 to v1.0.0-rc1 (master-935a1670-head). 

**1. Harvester pods did not crash -> `PASS`**
```
harvester-node01-upgrade100rc1:/home/rancher # kubectl get pods -n harvester-system
NAME                                                     READY   STATUS      RESTARTS   AGE
harvester-d544ddb6f-52mdk                                1/1     Running     0          58m
harvester-d544ddb6f-mg4dc                                1/1     Running     0          58m
harvester-d544ddb6f-npd7c                                1/1     Running     0          58m
harvester-load-balancer-59bf75f489-57nnp                 1/1     Running     0          58m
harvester-network-controller-68m6r                       1/1     Running     0          57m
harvester-network-controller-cdrjp                       1/1     Running     0          57m
harvester-network-controller-jfg96                       1/1     Running     0          58m
harvester-network-controller-manager-c57f8cbcb-67mcq     1/1     Running     0          58m
harvester-network-controller-manager-c57f8cbcb-9mqbx     1/1     Running     0          58m
harvester-network-controller-v2k2n                       1/1     Running     0          57m
harvester-node-disk-manager-29xrr                        1/1     Running     0          57m
harvester-node-disk-manager-v27hz                        1/1     Running     0          57m
harvester-node-disk-manager-wwpwf                        1/1     Running     0          58m
harvester-node-disk-manager-zwbxv                        1/1     Running     0          57m
harvester-promote-harvester-node02-upgrade100rc1-nphjz   0/1     Completed   0          9h
harvester-promote-harvester-node03-upgrade100rc1-72lj7   0/1     Completed   0          9h
harvester-webhook-67744f845f-pmrlg                       1/1     Running     0          57m
harvester-webhook-67744f845f-r5c44                       1/1     Running     0          58m
harvester-webhook-67744f845f-tqjkl                       1/1     Running     0          57m
kube-vip-2l2qp                                           1/1     Running     1          56m
kube-vip-cloud-provider-0                                1/1     Running     34         10h
kube-vip-cvklf                                           1/1     Running     0          56m
kube-vip-q99lt                                           1/1     Running     0          56m
virt-api-86455cdb7d-2hb4x                                1/1     Running     3          10h
virt-api-86455cdb7d-q8fpc                                1/1     Running     3          10h
virt-controller-5f649999dd-q5bqs                         1/1     Running     18         10h
virt-controller-5f649999dd-sqh9g                         1/1     Running     20         10h
virt-handler-4ncxn                                       1/1     Running     3          10h
virt-handler-cmzg2                                       1/1     Running     3          9h
virt-handler-k8pg9                                       1/1     Running     3          10h
virt-handler-x754t                                       1/1     Running     2          8h
virt-operator-56c5bdc7b8-cgwc8                           1/1     Running     28         10h
```

**2. Check whether longhornBackupName is in each VM Backup. -> `PASS`**

```
$ kubectl get backup -A
longhorn-system   backup-7933c0d09ec04d1a   snapshot-b1fbfcf4-ad45-442d-9bb6-8119e713d892   1367343104     2021-12-17T07:13:00Z   Completed   2021-12-17T07:17:51.097572079Z

```

```
$ kubectl get vmbackup ubuntu-backup -o yaml | less

volumeBackups:
  - creationTime: "2021-12-17T07:12:59Z"
    longhornBackupName: backup-7933c0d09ec04d1a
    name: ubuntu-backup-volume-ubuntu-vm-disk-0-ylodf
```

**3. Check whether there is <namespace>-<name>.cfg file in backup target. -> `PASS`**
  There is a default-ubuntu-backup.cfg file exists on S3 backup remote bucket 

``` $ cat default-ubuntu-backup.cfg ```

![image](https://user-images.githubusercontent.com/29251855/146564069-24d2f5d0-5b98-438d-b909-cf53ad2691ce.png)

**4. Check whether you can restore VM. -> `PASS`**
    Can create a new vm from restore the existing backup  

```
![image](https://user-images.githubusercontent.com/29251855/146564664-46809072-a320-44a6-8665-29aa1b8d936f.png)


```

**Environment:**
 - Harvester ISO version before upgrade: v0.3.0
 - Harvester ISO version after upgrade: v1.0.0-rc1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): 4 nodes harvester cluster on local kvm machines

**Harvester node information**
- VIP: 192.168.122.71
- node1: 192.168.122.253 (6 core, 12GB, 200GB)
- node2: 192.168.122.10 (6 core, 12GB, 200GB)
- node3: 192.168.122.36 (4 core, 12GB, 200GB)
- node4: 192.168.122.97 (4 core, 10GB, 200GB)

### Verify Steps
1. Download harvester v0.3.0 iso and do checksum
2. Download harvester v1.0.0 iso and do checksum
3. Use ISO Install a 4 nodes harvester cluster 
4. Create several OS images from URL
5. Create ssh key
6. Enable vlan network with `harvester-mgmt`
7. Create virtual network `vlan1` with id `1`
8. Create 2 virtual machines
- ubuntu-vm: 2 core, 4GB memory, 30GB disk

9. Setup backup target
10. Take a backup from ubuntu vm

**upgrade process**
Follow the manual upgrade steps to upgrade from v0.3.0 to v1.0.0-rc1
https://github.com/harvester/docs/pull/67/files

Add the following content to  `/usr/local/harvester-upgrade/upgrade-helpers/manifests/10-harvester.yaml` 
before upgrade harvester controller node

```
--
apiVersion: management.cattle.io/v3
kind: ManagedChart
metadata:
  name: harvester-crd
  namespace: fleet-local
spec:
  chart: harvester-crd
  releaseName: harvester-crd
  version: 0.0.0-dev
  defaultNamespace: harvester-system
  repoName: harvester-charts
  # takeOwnership will force apply this chart without checking ownership in labels and annotations.
  # https://github.com/rancher/fleet/blob/ce9c0d6c0a455d61e87c0f19df79d0ee11a89eeb/pkg/helmdeployer/deployer.go#L323
  # https://github.com/rancher/helm/blob/ee91a121e0aa301fcef2bfbc7184f96edd4b50f5/pkg/action/validate.go#L71-L76
  takeOwnership: true
  targets:
  - clusterName: local
    clusterSelector:
      matchExpressions:
      - key: provisioning.cattle.io/unmanaged-system-agent
        operator: DoesNotExist
  values: {}
```

### Additional Context

Since issue #1645, currently we can't access harvester dashboard by VIP
And since #1666, after upgrade we are still not able to login with original admin password
